src/common/config_parser.h

```diff
@@ -16,48 +13,74 @@
 
 namespace marian {
 
-enum struct ConfigMode {
-  training,
-  translating,
-  rescoring,
-};
-
-// try to determine the width of the terminal
-uint16_t guess_terminal_width(uint16_t max_width = 180);
+namespace cli {
+enum struct mode { training, translation, scoring };
+}  // namespace cli
 
+/**
+ * @brief Command-line options parser
+ *
+ * New options and aliases should be defined within `addOptions*` methods.
+ */
 class ConfigParser {
 public:
-  ConfigParser(int argc, char** argv, ConfigMode mode, bool validate = false)
-      : mode_(mode),
-        cmdline_options_("Allowed options", guess_terminal_width()) {
+  ConfigParser(int argc, char** argv, cli::mode mode, bool validate = false)
+      : mode_(mode) {
     parseOptions(argc, argv, validate);
   }
 
+  /**
+   * @brief Parse command-line options
+   *
+   * Options are parsed in the following order, later config options overwrite
+   * earlier:
+   *  * predefined default values
+   *  * options from the config files provided with --config, from left to right
```

Comment:
Does still work for interpolating environment variables? (Maybe could you add a test case? This is critical for our production training.)

Reply:
I was pretty sure I've added a test case for that, but I'll check.

Reply:
Added a basic test here: https://github.com/marian-nmt/marian-regression-tests/blob/master/tests/interface/envvars/test_interpolate_envvars.sh